### ISSUE TITLE
Line changes result in incorrect selections

### ISSUE THREAD

**Issue by aMoniker** [desktop/desktop#4720]
## Description
When you've got a few lines selected, then make a change to the file which adds or subtracts lines that come before your selections, the selections will remain, but they'll be on the wrong lines. They probably only track line number, and fail to update when lines change.

## Version
* GitHub Desktop: 1.1.1
* Operating system: OSX 10.13.4

## Steps to Reproduce
1. Make changes to a file
2. Select some of the changed lines in GH Desktop, but don't commit
3. Add or subtract some lines in the file that fall before your selected lines
4. See that your selected lines are now different in GH Desktop

### Expected Behavior
It may not be possible to intelligently guess which lines should be selected after changes were applied. In the case where the selection cannot be reliably determined, simply clear that selection. In all other cases where it can be determined, update the selection to the new and correct line.

### Actual Behavior
Selections remain, but for the wrong lines.

**Comment by iAmWillShepherd** [comment]
Thanks for taking the time to open an issue! 🙇 

I was able to reproduce this and have marked this as a bug.

![repro](https://user-images.githubusercontent.com/1715082/40191383-01bfcc26-59c7-11e8-9f86-29e184760dfe.gif)

**Comment by shiftkey** [comment]
> It may not be possible to intelligently guess which lines should be selected after changes were applied. In the case where the selection cannot be reliably determined, simply clear that selection.

Yeah, I've been involved with implementing partial selection twice now and I think reliably "moving" the selection when line contents change is really hard to do reliably.

I'm going to move this to `enhancement` because, while the line selections don't match up with the content as it moves, they are consistent with the overall file length (e.g. line 7 remains selected when the content changes).

**Comment by iAmWillShepherd** [comment]
@shiftkey, how do you feel about clearing the selection if the file changes. Doing so could help people avoid committing the wrong thing.

**Comment by shiftkey** [comment]
@iAmWillShepherd I feel that would add a lot more manual work to crafting partial commits if it's tied to the current file contents.

One example that comes to mind:

 - user selects lots of lines they want to commit
 - user notices a typo in one line, switches back to editor
 - switches back to the app, line selections are remembered with the right content shown

With this hypothetical change:

 - user selects lots of lines they want to commit
 - user notices a typo in one line, switches back to editor
 - switches back to the app, no lines are selected and the user has to re-select all the lines

**Comment by j-f1** [comment]
How about deselecting ant lines that are changed after being selected?

**Comment by shiftkey** [comment]
@j-f1 I still think it's the same underlying problem - detecting when the content of a line has changed from what it was when you selected it.

**Comment by shiftkey** [comment]
@patricktokeeffe opened #4843 which overlaps with this, but added this comment:

> in my testing with v1.2.2, it seems the selections are only be modified wrt the changeblock that appears directly after the edited line. I did not observe line selections changing if no edit occurred between a block and its prior.
